File: jetbrains-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/Situation.kt
package com.ezevillo.gitreview.domain

/**
 * Outcome of an invocation (data-model.md § Situation). Pure logic, no IDE.
 */
enum class Situation {
    REVIEW,
    NO_REVIEW,
    OUT_OF_RANGE,
    ERROR,
    CLI_MISSING,
    CLI_OUTDATED,
    FINISH_CONFLICT,
    FINISH_PENDING,
    ;

    /** Wire / UI id matching the VS Code extension (`review`, `cli-missing`, …). */
    val id: String
        get() = when (this) {
            REVIEW -> "review"
            NO_REVIEW -> "no-review"
            OUT_OF_RANGE -> "out-of-range"
            ERROR -> "error"
            CLI_MISSING -> "cli-missing"
            CLI_OUTDATED -> "cli-outdated"
            FINISH_CONFLICT -> "finish-conflict"
            FINISH_PENDING -> "finish-pending"
        }

    companion object {
        fun fromId(id: String): Situation? = entries.find { it.id == id }
    }
}

/**
 * Maps `status --porcelain` exit code to [Situation]. Unknown codes (including 1)
 * are always [Situation.ERROR], never [Situation.REVIEW].
 */
fun situationForExitCode(exitCode: Int?): Situation = when (exitCode) {
    0 -> Situation.REVIEW
    2 -> Situation.NO_REVIEW
    3 -> Situation.OUT_OF_RANGE
    else -> Situation.ERROR
}

/**
 * Extends [situationForExitCode] with finish records:
 * `finish-conflict` over `review`, `finish-pending` over `no-review`.
 */
fun situationFor(
    exitCode: Int?,
    hasFinishConflict: Boolean,
    hasFinishPending: Boolean,
): Situation {
    val base = situationForExitCode(exitCode)
    if (base == Situation.REVIEW && hasFinishConflict) return Situation.FINISH_CONFLICT
    if (base == Situation.NO_REVIEW && hasFinishPending) return Situation.FINISH_PENDING
    return base
}

/**
 * Situations where [ReviewState] has a populated state record and read/open
 * commands are safe: active review or finish-conflict.
 */
fun isReviewReadable(situation: Situation): Boolean =
    situation == Situation.REVIEW || situation == Situation.FINISH_CONFLICT

File: visualstudio-extension/src/GitReview.Domain/Situation.cs
namespace GitReview.Domain;

/// <summary>Outcome of an invocation (data-model § Situation). Pure logic, no IDE.</summary>
public enum Situation
{
    Review,
    NoReview,
    OutOfRange,
    Error,
    CliMissing,
    CliOutdated,
    FinishConflict,
    FinishPending,
}

public static class SituationIds
{
    public static string Id(this Situation s) => s switch
    {
        Situation.Review => "review",
        Situation.NoReview => "no-review",
        Situation.OutOfRange => "out-of-range",
        Situation.Error => "error",
        Situation.CliMissing => "cli-missing",
        Situation.CliOutdated => "cli-outdated",
        Situation.FinishConflict => "finish-conflict",
        Situation.FinishPending => "finish-pending",
        _ => throw new ArgumentOutOfRangeException(nameof(s)),
    };

    public static Situation? FromId(string id) => id switch
    {
        "review" => Situation.Review,
        "no-review" => Situation.NoReview,
        "out-of-range" => Situation.OutOfRange,
        "error" => Situation.Error,
        "cli-missing" => Situation.CliMissing,
        "cli-outdated" => Situation.CliOutdated,
        "finish-conflict" => Situation.FinishConflict,
        "finish-pending" => Situation.FinishPending,
        _ => null,
    };

    /// <summary>
    /// Maps <c>status --porcelain</c> exit code to Situation.
    /// Unknown codes (including 1) are always Error, never Review.
    /// </summary>
    public static Situation ForExitCode(int? exitCode) => exitCode switch
    {
        0 => Situation.Review,
        2 => Situation.NoReview,
        3 => Situation.OutOfRange,
        _ => Situation.Error,
    };

    /// <summary>
    /// Extends <see cref="ForExitCode"/> with finish records:
    /// finish-conflict over review, finish-pending over no-review.
    /// </summary>
    public static Situation For(
        int? exitCode,
        bool hasFinishConflict,
        bool hasFinishPending)
    {
        var bas = ForExitCode(exitCode);
        if (bas == Situation.Review && hasFinishConflict) return Situation.FinishConflict;
        if (bas == Situation.NoReview && hasFinishPending) return Situation.FinishPending;
        return bas;
    }

    public static bool IsReviewReadable(Situation situation) =>
        situation is Situation.Review or Situation.FinishConflict;
}
